### ISSUE TITLE
Is the second softmax application necessary?

### ISSUE THREAD

**Issue by maxdinech** [bethgelab/foolbox#158]
In [`criteria.py`](https://github.com/bethgelab/foolbox/blob/master/foolbox/criteria.py), the `predictions` parameters of `.is_adversarial()` is described as "A vector with the pre-softmax predictions for some image" (line  #96).

This is troublesome : in 99% of the cases, a classifier network's last layer activation function already is a softmax, thus applying softmax again to values that already are probabilities (lines 282 and 322), and completely changing those probabilities.

This should at least be documented (it took me a long time to figure out why the `.is_adversarial()` function behaved strangely), or possibly even changed: there could be a `pre_softmax=True or False` argument in the criterion class to manage both cases.

I am ready to make a pull request to implement this, do you think that it could be a good idea?

**Comment by jonasrauber** [comment]
I guess it depends on what framework you use. For TensorFlow, this is properly [documented](https://foolbox.readthedocs.io/en/latest/modules/models.html#foolbox.models.TensorFlowModel): 

> The predictions of the model, before the softmax.

For Keras, we actually default to probabilities and allow the user to specify what's passed using the `predicts` argument (see [here](https://foolbox.readthedocs.io/en/latest/modules/models.html#foolbox.models.KerasModel)):

> Specifies whether the Keras model predicts logits or probabilities. Logits are preferred, but probabilities are the default.

The problem with supporting probabilities is that calculating gradients based on the probability output is numerically unstable.

If you have recommendations where we can improve the documentation that would be great. Also, which framework are you using?

**Comment by maxdinech** [comment]
I am using Pytorch.

For my project I just removed the softmax application in the foolbox library code and everything went well : I obtained the expected results.

I didn't mean to criticize the documentation, but merely that the tutorial should point this out, as this default behavior is (at least to me) really not obvious.

Also, for projects in which the attacked network is not editable, there seems to be no other way than modifying the library code to get proper results. Maybe this case is not that frequent, but I think that is those cases adding an argument in the foolbox model class specifying is the network already applies a softmax could be useful.

Thanks a lot anyway for this wonderful library !

**Comment by jonasrauber** [comment]
So you removed the softmax in criteria and your model is returning probabilities directly? The problem with that is that gradient calculation will be wrong in that case. It might not matter for decision-based attacks, but the remaining attacks will not work correctly when your model returns probabilities instead of logits.
In other words: it's not as simple as removing the softmax in the criteria; to calculate and differentiate the cross-entropy in a numerically stable way, the logits are needed.

In Keras and TensorFlow it's relatively easy to access the logits even if the model outputs probabilities without having to modify the model (and in fact for Keras we even have built-in support for this because there it's common to have a model that outputs probabilities), but I am not sure if this is possible in PyTorch.
If this (i.e. accessing the input tensor of another tensor, e.g. traversing the graph backwards from the probabilities-tensor to the logits tensor) is possible in PyTorch, it would be a great addition to properly support your use-case.

**Comment by wielandbrendel** [comment]
One could in fact check the last layer whether it has an activation of type 'softmax' or if it is a softmax non-linearity (which could then be removed). Shall we add this on our to do list, @jonasrauber . We should make a one-day hackathon at some point to add all these little features here and there.

**Comment by jonasrauber** [comment]
@wielandbrendel absolutely, if there is a way in PyTorch to get the input tensor of the softmax when you are only given its output Tensor, we should add a more automatic way to do this… I am just not sure whether this is easily possible in PyTorch using a public API
